<div style="float: right; color: red;">Please, rename this file to <code>HW5.ipynb</code> and save it in <code>MSA8010F16/HW5</code>
</div>

# Homework 5: Evaluating Classifiers

In [33]:
import matplotlib as plt
import numpy as np

In [2]:
from sklearn import datasets
iris = datasets.load_iris()


In [3]:
iris.data[0:4]

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2]])

In [4]:
iris.target[0:4]

array([0, 0, 0, 0])

In [5]:
from sklearn import svm
clf = svm.SVC()     ## create our classifier

In [23]:
X = iris.data
y = iris.target
clf.fit(X, y)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [7]:
y_hat = clf.predict(X)

In [9]:
sum(y_hat - y)

2

#### Our algorithm from class had only two misclassifications out of 150, or 1.3% failure rate.

### First attempt

I found that the scikit-learn package includes a function to generate test datasets for a variety of purposes. I will try to generate a test classification dataset and run the algorithm on it.

In [25]:
x2,y2 = datasets.make_classification(n_samples=2000, n_features=15, n_informative=10, n_classes=5)

In [27]:
x2[0:4]

array([[ 2.18324371,  1.1223366 ,  2.25092446,  0.53728165,  0.39556735,
         4.87397706,  1.41861617,  1.51372194, -2.11305509, -0.58453928,
         1.39574229,  3.0722766 , -0.68290405,  2.32114064, -0.43820605],
       [-0.07485471,  3.96727078,  3.14192513,  1.37319613,  0.59709896,
        -2.35092509, -1.70107552, -1.44885827,  5.01218273, -0.59532572,
        -1.68578544, -0.27801341, -0.16372603, -1.37878229, -0.60380773],
       [-0.95307446,  2.99678339, -1.43705898, -0.6223117 , -0.81818878,
         0.15955234, -0.99116889,  1.21850853, -3.53449297, -0.34586332,
        -0.80972945,  2.16991343, -2.31099923,  0.08711621, -2.44412046],
       [ 2.13416994, -1.34178984, -0.25497998, -0.57843472, -0.14906444,
        -1.26445084,  1.69914568,  0.12643926,  0.34876414,  0.2966645 ,
        -1.17926504,  0.4333565 , -1.3375751 ,  0.25263622,  1.60774043]])

In [28]:
y2[0:4]

array([4, 0, 3, 1])

In [29]:
clf2 = svm.SVC()
clf2.fit(x2, y2)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [30]:
y2_hat = clf2.predict(x2)

In [35]:
np.count_nonzero(y2 == y2_hat)

1945

In [40]:
1945./2000.

0.9725

Our classifier model was able to properly classify 1945 of the 2000 generated samples, meaning 55 misclassifications. This is a 97.25% success rate.

### Attempt two - trying a real dataset

There are many available example datasets, including many geared towards classification machine learning. The challenge will be to figure out how to implement these into the scikit-learn algorithm.

I've found a set that gives the levels of 9 different elements in 7 types of glass.

I'll attempt to load the features as an array into a variable and the classes into their own array.

In [55]:
x3 = np.genfromtxt('glass.csv', delimiter=',', usecols=(range(9)))
y3 = np.genfromtxt('glass.csv', delimiter=',', usecols=(9))

In [56]:
x3[0:4]

array([[  1.51588824,  12.87795   ,   3.43036   ,   1.40066   ,
         73.282     ,   0.68931   ,   8.04468   ,   0.        ,   0.1224    ],
       [  1.5176423 ,  12.9777    ,   3.53812   ,   1.21127   ,
         73.002     ,   0.65205   ,   8.52888   ,   0.        ,   0.        ],
       [  1.52212996,  14.20795   ,   3.82099   ,   0.46976   ,
         71.77      ,   0.11178   ,   9.5726    ,   0.        ,   0.        ],
       [  1.52222108,  13.21045   ,   3.7716    ,   0.79076   ,
         71.9884    ,   0.13041   ,  10.2452    ,   0.        ,   0.        ]])

In [57]:
y3[0:4]

array([ 1.,  1.,  1.,  1.])

In [58]:
clf3 = svm.SVC()
clf3.fit(x3,y3)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [59]:
y3_hat = clf3.predict(x3)

In [60]:
np.count_nonzero(y3 == y3_hat)

162

In [61]:
np.size(y3)

214

In [62]:
162. / 214.

0.7570093457943925

This time, the algorithm was only able to correctly identify 75.7% of the 214 samples of glass.

### Attempt three - one more example

Another interesting dataset I found was a classifcation of three wine types based on 13 numerically measurable characteristics. Let's again try to load the proper values into two arrays to run our classification algorithm on.

In [95]:
y4 = np.genfromtxt('wine.data.txt', dtype=int, delimiter=',', usecols=(0))
x4 = np.genfromtxt('wine.data.txt', delimiter=',', usecols=(range(1,14)))

In [97]:
x4[0:2]

array([[  1.42300000e+01,   1.71000000e+00,   2.43000000e+00,
          1.56000000e+01,   1.27000000e+02,   2.80000000e+00,
          3.06000000e+00,   2.80000000e-01,   2.29000000e+00,
          5.64000000e+00,   1.04000000e+00,   3.92000000e+00,
          1.06500000e+03],
       [  1.32000000e+01,   1.78000000e+00,   2.14000000e+00,
          1.12000000e+01,   1.00000000e+02,   2.65000000e+00,
          2.76000000e+00,   2.60000000e-01,   1.28000000e+00,
          4.38000000e+00,   1.05000000e+00,   3.40000000e+00,
          1.05000000e+03]])

In [98]:
y4[0:2]

array([1, 1])

In [99]:
clf4 = svm.SVC()
clf4.fit(x4,y4)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [100]:
y4_hat = clf4.predict(x4)

In [103]:
np.count_nonzero(y4 == y4_hat)

178

In [102]:
np.size(y4)

178

This time, the classifier was 100% accurate. 